### PR TITLE
Add metrics and do no throttle requests on user quota when quota throttling mode is set as TRACKING.

### PR DESCRIPTION
**Pull request by ankagrawal**

### CODE REVIEW

File: ambry-frontend/src/main/java/com/github/ambry/frontend/AmbrySecurityService.java [hunk 1]
```diff
@@ -138,9 +139,11 @@ public void postProcessRequest(RestRequest restRequest, Callback<Void> callback)
         if (quotaManager != null) {
           ThrottlingRecommendation throttlingRecommendation = quotaManager.getThrottleRecommendation(restRequest);
           if (throttlingRecommendation != null && throttlingRecommendation.shouldThrottle()) {
-            Map<String, String> quotaHeaderMap = RestUtils.buildUserQuotaHeadersMap(throttlingRecommendation);
-            throw new RestServiceException("User Quota Exceeded", RestServiceErrorCode.TooManyRequests, true, true,
+            if (quotaManager.getQuotaConfig().throttlingMode == QuotaMode.THROTTLING) {
```
Comment (justinlin-linkedin) [inline ambry-frontend/src/main/java/com/github/ambry/frontend/AmbrySecurityService.java:142]: can we move this logic to `shouldThrottle` method in ThrottlingRecommendation? then we don't need two layers of if statement.

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java [hunk 2]
```diff
@@ -57,36 +62,52 @@ public AmbryQuotaManager(QuotaConfig quotaConfig, ThrottlePolicy throttlePolicy,
...
+    Timer.Context timer = quotaMetrics.quotaManagerInitTime.time();
     for (QuotaEnforcer quotaEnforcer : requestQuotaEnforcers) {
       quotaEnforcer.init();
     }
+    timer.stop();
   }
 
   @Override
   public ThrottlingRecommendation getThrottleRecommendation(RestRequest restRequest) {
+    Timer.Context timer = quotaMetrics.quotaEnforcementTime.time();
```
Comment (justinlin-linkedin) [inline ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java:79]: please move this line after if statement below.

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java [hunk 3]
```diff
@@ -101,6 +122,11 @@ public QuotaConfig getQuotaConfig() {
     return quotaConfig;
   }
 
+  @Override
+  public QuotaMetrics getQuotaMetrics() {
```
Comment (justinlin-linkedin) [inline ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java:126]: doesn't seem like it's used anywhere?

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java (hunk 2 above)
Comment (ankagrawal) [reply ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java:79]: done.

File: ambry-frontend/src/main/java/com/github/ambry/frontend/AmbrySecurityService.java (hunk 1 above)
Comment (ankagrawal) [reply ambry-frontend/src/main/java/com/github/ambry/frontend/AmbrySecurityService.java:142]: I have changed the logic to remove the layer of if statement. We don't really need to separate `if`.

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java (hunk 3 above)
Comment (ankagrawal) [reply ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java:126]: removed.

File: ambry-api/src/main/java/com/github/ambry/quota/QuotaMetrics.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.Timer;
+
+
+/**
+ * Metrics class to capture metrics for user quota enforcement.
+ */
+public class QuotaMetrics {
```
Comment (justinlin-linkedin) [inline ambry-api/src/main/java/com/github/ambry/quota/QuotaMetrics.java:24]: we probably shouldn't put QuotaMetrics here in the ambry-api subproject. It's not used by another other subproject outside of ambry-quota. I recommend putting it under ambry-quota subproject. And maybe inside AmbryQuotaManager class as an inner class.

Comment (ankagrawal) [reply ambry-api/src/main/java/com/github/ambry/quota/QuotaMetrics.java]: fixed.

File: ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java [hunk 5]
```diff
@@ -156,4 +187,27 @@ public QuotaConfig getQuotaConfig() {
     }
     return quotaSourceObjectMap;
   }
+
+  /**
+   * Metrics class to capture metrics for user quota enforcement.
+   */
+  public class QuotaMetrics {
```
Comment (justinlin-linkedin) [inline ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java:194]: make it private and static please.

**Review by justinlin-linkedin — APPROVED** [review]

**Review by zzmao — APPROVED** [review]
LGTM

**Pull request merged**